### PR TITLE
Resolve Instance `Elector` from HTML response

### LINKED ISSUE #1: Instance `Elector` from HTML response
## Requirement

The `From` trait which is already implemented for the `Elector` struct should return an `Elector` instance given the HTML document returned by:
`http_client.get(&url).send().await?`

### `Elector` `From` trait Implementation

https://github.com/rust-lang-ve/libcne/blob/8461bc153bcbaf3448eb8ada2a06840e625a97d3/src/cne/elector.rs#L20

### Request that returns HTML

https://github.com/rust-lang-ve/libcne/blob/8461bc153bcbaf3448eb8ada2a06840e625a97d3/src/request/request.rs#L11

### PR DESCRIPTION
**Pull request by angel-afonso**
Implemented `From` trait for `Elector`

### CODE REVIEW

File: example/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,3 @@
+fn main() {
+    println!("Hello, world!");
```
Comment (LeoBorai) [inline example/src/main.rs]: Why adding an example that calls `Hello World`?

File: html.html [hunk 2]
```diff
@@ -0,0 +1,168 @@
+<!--<a href="../../registro_electoral/impugnacionrenovacion.pdf" target="_blank"> DESCARGA LA PLANILLA</a>-->
```
Comment (LeoBorai) [inline html.html]: Please, remove files that are not related to the crate, if you need this content to test your code, leave it inside an string in your tests.

File: example/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+[package]
```
Comment (LeoBorai) [inline example/Cargo.toml]: Why are we using an empty example?

**Review by LeoBorai — CHANGES_REQUESTED** [review]

File: src/cne/elector.rs [hunk 4]
```diff
@@ -17,6 +18,236 @@ pub struct Elector {
 
 impl From<String> for Elector {
   fn from(html_string: String) -> Self {
-    todo!()
+    let document = Html::parse_fragment(&html_string);
+    let selector = Selector::parse("table tbody tr td table tbody tr:last-child td table tbody tr:nth-child(2) td table:first-child tbody tr").unwrap();
```
Comment (LeoBorai) [inline src/cne/elector.rs:37]: I think we could improve this by sanitizing the data before processing it.
As we now the response is a template, we could assume that for every response (of a successful search), between lines 1 and 43 (from your `html.html` file), there's no resources of our interest in order to build an `Elector`.

Then we could ignore those lines at first, that will help by making this `selector` much easier to use.

A good starting point could be:

```rust
    fn sanitize_html(text: String) -> String {
      let mut sanitized = String::default();

      text.split("\n").into_iter().enumerate().for_each(|(i, x)| {
        if i > 43 {
          sanitized.push_str(x);
        }
      });

      sanitized
    }
```

File: src/cne/elector.rs [hunk 5]
```diff
@@ -17,6 +18,236 @@ pub struct Elector {
...
+        .next()
+        .unwrap()
+        .inner_html(),
+      parish: select
+        .next()
+        .unwrap()
+        .select(&field_selector)
+        .next()
+        .unwrap()
+        .inner_html(),
```
Comment (LeoBorai) [inline src/cne/elector.rs]: Is there another way to accomplish this? I can see the same code many times, why not writing a function to make it more readable and more "DRY"?

**Review by LeoBorai — CHANGES_REQUESTED** [review]
Review on HTML parser

File: src/cne/elector.rs (hunk 5 above)
Comment (angel-afonso) [reply src/cne/elector.rs]: I already refactored the code by dividing the selection into 3 functions

File: html.html (hunk 2 above)
Comment (angel-afonso) [reply html.html]: Sorry it was a mistake i already deleted those files

File: src/cne/elector.rs [hunk 6]
```diff
@@ -17,6 +39,316 @@ pub struct Elector {
 
 impl From<String> for Elector {
   fn from(html_string: String) -> Self {
-    todo!()
+
+    if verify_citizenship_error(&html_string) {
+      panic!("Invalid citizenship provided");
```
Comment (LeoBorai) [inline src/cne/elector.rs]: This doesn't goes here, the `From` trait is not responsible for validating the provided value.
Before trying to create an `Elector` from an arbitrary value, first you must check if the provided HTML is a valid candidate.

I would move this validation to the `request` module, so I could check if the response is valid before getting here.

Comment (LeoBorai) [reply src/cne/elector.rs]: Also notice that the `from` function signature from the `From` trait always returns `Self`, this means that there's no way to return `Result` and handle the error from the caller side.

Now imagine that your app is running on production and it suddenly panics (causing your application to stop its execution), just because an invalid search has been made, I think you could recover from that easily!

File: src/cne/elector.rs [hunk 7]
```diff
@@ -1,5 +1,27 @@
...
+  let selector = Selector::parse("b").unwrap();
+  field.select(&selector).next().unwrap()
+}
+
+fn select_font_tag<'a>(field: &'a mut ElementRef<'a>) -> scraper::element_ref::ElementRef<'a> {
+  let selector = Selector::parse("font").unwrap();
+  field.select(&selector).next().unwrap()
+}
+
+fn verify_citizenship_error(html_string: &str) -> bool {
```
Comment (LeoBorai) [inline src/cne/elector.rs]: This should be validated before getting here. Is the request responsibility to check if the response is successful or not, `Elector` should be responsible of handling `Elector` related logic, this validation belongs to the request itself.

**Review by LeoBorai — CHANGES_REQUESTED** [review]

File: src/cne/elector.rs [hunk 8]
```diff
@@ -17,6 +39,316 @@ pub struct Elector {
...
+      state: select_field(&mut selection).inner_html(),
+      town: select_field(&mut selection).inner_html(),
+      parish: select_field(&mut selection).inner_html(),
+      voting_center: select_font_tag(&mut select_field(&mut selection)).inner_html(),
+      address: select_font_tag(&mut select_field(&mut selection)).inner_html(),
+    }
+  }
+}
+
+#[cfg(test)]
```
Comment (LeoBorai) [inline src/cne/elector.rs]: Test are now set on the `master` branch, we could move this to the tests directory

**Comment by LeoBorai** [comment]
Hey @angel-afonso !
I canceled the workflows because the test task has been running for an hour and a half!
Looks like some deadlock happened or something similar!

As we are using `reqwest` crate as an HTTP client, we could set a timeout for the request, just in case the host doesn't respond.

Given the size of the payload and the average time of response i think that 10s is good enough for the worst case scenario.

Give a shot on this: [Reqwest Client Timeout](https://docs.rs/reqwest/0.10.7/reqwest/struct.ClientBuilder.html#method.timeout), maybe this will help!

PD: It would be nice to have a configurable timeout with a default!

**Review by LeoBorai — APPROVED** [review]
Excellent! Thanks!
